### ISSUE TITLE
[extension types] ability to hide all safe "escape hatches" within a library?

### ISSUE THREAD

**Issue by modulovalue** [dart-lang/language#3365]
Consider the following program:

```dart
void main() {
  final foo = Foo(0);
  print(foo._i);
}

extension type Foo(int _i) {}
```

Is it possible to hide `_i` from all declarations declared in the same library that Foo has been declared in? In other words, to restrict Foo in a way that only a cast (`foo as int`) can be used to extract the underlying value?

I would like to make `print(foo._i);` a compile-time error so that all access to the underlying value has to be explicitly provided by `Foo`, but it seems like that is not possible. I would like to use such a more restrictive "policy" to prevent accidental access to the underlying value.

**Comment by lrhn** [comment]
Not currently possible. Not likely that we make it possible. Hiding something from the same library is generally not a goal.

If you do want the private name to be hidden, you can put it in a separste library.
It's assumed that if something is inside the same library, it's allowed to depend on the concrete implementation, including what the representation object getter is named.

Whether something inside the same library writes `x._d` or `x as int` makes no difference, it does exactly the same thing (which is "nothing" except changing the static type).
So, what are you trying to protect yourself from?

You can make the name even less attractive to use, say by making it 
```dart
extension type Foo(int _) {}
```
Or
```dart
extension type Foo(int _$dontUseThisName) {}
```

**Comment by modulovalue** [comment]
@lrhn said:

> Whether something inside the same library writes x._d or x as int makes no difference, it does exactly the same thing 

In my view, casting is definitely different and "unsafe" as it escapes the "soundness" guarantees of the type-system and can lead to accidental `TypeError`s at runtime. Both ways of accessing the underlying value are similar, but they are not the same:

`int i = foo._i;` says: _Only compile successfully if \_i can be assigned to a value of type int_.
`int i = foo as int;` says: _I don't care what \_i is, even if it isn't an int, I want the program to compile successfully, feel free to crash at runtime if I'm wrong_.


This program compiles successfully, as expected.

```dart
void foo(Foo foo) {
  final int access = foo._i;
  print(access);
  final int cast = foo as int;
  print(cast);
}

extension type Foo(int _i) {}
```

However, if the type of `Foo._i` is changed to, e.g., a String, the cast introduces a nasty runtime exception that static analysis doesn't protect us from:

```dart
void foo(Foo foo) {
  // Error: A value of type 'String' can't be assigned to a variable of 
  // type 'int'.
  // final int access = foo._i;
  //                        ^
  final int access = foo._i;
  print(access);
  // No static warnings (therefore, unsafe).
  final int cast = foo as int;
  print(cast);
}

extension type Foo(String _i) {}
```

This is what I was referring to by "unsafe" (_unfortunately, I'm not aware of a less overloaded term for this observation_).

---

> You can make the name even less attractive to use than [...]

Fair point, I agree, but it still "feels" somewhat suboptimal to make the default be so permissive and not give users a way to be more restrictive. 

Imagine, for example, _in the distant future_, a tool that auto-completes your program based on type information. An extension type could be used to enforce strict "access control" on a value, but if we can't hide the underlying value, extension types become a mere "suggestion" rather than a tool for sound access control (at least in the library that they are declared in).

What about a feature that mirrors `extensions` by supporting an on clause and an implicit one argument constructor?

```dart
void main() {
  int i = Foo(0);
  // There's no member that exposes the underlying value.
  Bar j = Bar(0);
}

extension Foo on int {}

// The underlying value can be accessed via `this`.
extension type Bar on int {}
```

Users should already be familiar with on clauses from extensions, so supporting them on extension types would feel natural, and would relieve us from the burden of having to give the underlying value a name.

**Comment by lrhn** [comment]
A cast is definitely more error-prone than a typed member access. It's effectively the same when the type is correct.

Allowing things inside the same library, that we don't allow outside, is not new to this. It's the same for class modifiers (you can implement a `final` class inside the same library, there is no way to prevent that) and private members (privacy is library privacy). 
The language generally does not try to prevent code in the same library from doing anything that isn't unsound. Access and feature restrictions are for other people, code in the same library is assumed to be trusted implicitly, and to be *aware* of the implementation, not just the interfaces. 
If there is something that code shouldn't do, it simply doesn't do it. The author of the library is in complete control, and we assume everything they do is deliberate.

Which means that "protecting the library against itself" is unlikely to be enough reason to add complications to the design for extension types. 
Protecting the library against other libraries, with possibly other authors that don't know the rules and implementation choices, is good. I try to keep the abstraction line clear, so authors can know what their clients can see and do, and what they cannot see or depend on.
Code in the same library is inside the abstraction. Whatever it does should be cooperative, not treated as adversarial.


In this case, just allowing you to not have a getter is probably possible. I'd go with `Foo(int)` with no name for that, not try to invent a secondary syntax. But it's still an extra complication with no real benefit over `Foo(int _)`, if protecting against the same library is not considered a benefit. Which it isn't.

**Comment by modulovalue** [comment]
> Allowing things inside the same library, that we don't allow outside, is not new to this.

Of course, but this argument can also be made the other way around, that is, forbidding things inside the same library is also not new. For example, constructors, `this` and `super` of a Foo are not visible to other declarations inside of the same library from an instance of Foo.

```dart
class Foo {
  Foo();
  
  Foo.bar();
}

void main() {
  Foo().this; // The getter 'this' isn't defined for the type 'Foo'.
  Foo().super; // The getter 'super' isn't defined for the type 'Foo'.
  Foo().bar(); // The method 'bar' isn't defined for the type 'Foo'.
}
```

Therefore, I don't think that it makes much sense to compare the visibility-related aspects of features across features since there isn't one consistent way of dealing with visibility across all of them.

---

> I'd go with Foo(int) with no name for that, not try to invent a secondary syntax.

`Foo(int)` would also be new syntax, wouldn't it? If that is something that the primary constructor proposal is intending to support, then I would agree with that. However, as I've already mentioned, users are already familiar with on clauses from extensions, and it seems that `this` is already bound to the represented value, so adding an on clause as an alternative to the primary constructor doesn't seem like a huge complication: 

1. It is a compile-time error if an extension type declares a primary constructor and an on clause at the same time.
2. If an extension type declares an on clause, it's represented type is the type that the on clause specifies.
3. If an extension type declares an on clause `on X`, then an implicit primary constructor `(X $)` is introduced where `$` is a fresh identifier that is not exposed through the scope of the extension type.


It looks to me like that would suffice to specify on clauses.

> with no real benefit over Foo(int _)

I have to disagree and I do see real benefits, to be clear:

- specifying an unreadable name for the represented value (e.g., `extension type Foo(int _$dontUseThisName) {}`) pollutes the scope of the extension type with an identifier that clearly is not meant to be there.
- a redundant name is introduced into the scope of Foo because the represented value can already be accessed via `this`. 
- a primary constructor forces users to give the represented value a name, and it's unclear to me that, in general, clear and useful names do exist here (`value`, `self`, `_`?).


@eernstg since you are the primary author of the [extension types](https://github.com/dart-lang/language/issues/2727) [proposal](https://github.com/dart-lang/language/blob/main/accepted/future-releases/extension-types/feature-specification.md), I'd like to ask you for some feedback on this issue.

**Comment by modulovalue** [comment]
I have to disagree with myself on the on clause. While I do think that it's syntax that users are already familiar with, what I've proposed wouldn't support making the primary constructor private.


> [...] We could, say by allowing extension type Foo(int) with no name for the representation object

Because syntax like this could support making the primary constructor private and hide the representation object from the library that Foo was declared in, this seems more functional than an on clause.

However, I do still believe that supporting this would have benefits, some of which I've enumerated in the previous comment.

**Comment by eernstg** [comment]
I tend to agree with @lrhn that there is little justification for having a mechanism that prevents access to a declaration in the same library (including parts, if any): A simple textual search or an IDE reverse lookup will reveal any such reference, and you can edit the code if any of those usages is in some sense "wrong". 

(OK, some occurrences of `_i` could refer to some other declaration, so you may get too many hits with a textual search, but the main point is that you won't miss any.)

If you wish to make it an error to access the representation variable even in the same library, you could consider using the type `void`:

```dart
extension type DontTouchMyRepresentation(void _v) {
  void foo() {
    print('Members should never access the representation object, but '
        'they can of course still do other things, like a static member');
  }
}

void main() {
  var x = DontTouchMyRepresentation(1);
  print(x); // Oh yes, we can still print anything.
  Object? o = x._v; // But this is a compile-time error.
}
```

This gets close to the behavior you'd have for a representation object that has no name. In particular, it yields a compile-time error for all accesses to the representation variable in the same library, and privacy will make the representation variable appear to be nameless outside of the library (or you could use a public name and get the `void` treatment in all libraries).

We have had discussions about a different kind of access control that I'd consider to be more useful: An instance member declaration could be declared to be available using the lexical scope only.

Let's say that we get that treatment by saying in the declaration that the member is `lexical` (that's not a serious proposal, just a device that allows us to talk about it here).

```dart
class A {
  lexical int i;
  A(this.i); // OK.
  void foo() => print(i); // OK
  void bar() => print(this.i); // Compile-time error.
}

class B extends A {
  B(super.i); // OK, this is just a parameter declaration.
  void bar(A other) {
    print(i); // Compile-time error, there is no `i` in the lexical scope.
    print(other.i); // Compile-time error, there is no `i` in the interface of `A`.
  }
}

var _ = B(1).i; // Compile-time error, there's no `i` in the interface of `B`.
```

This concept would be applicable to instance member declarations in various different kinds of enclosing declarations (in classes, in mixins, in enums, in extension types, in extensions, plus whatever we'll come up with in the future).

One way to explain how this works could be to say that it's like an auto-renamed instance member declaration: `lexical int i;` really means something like `int _$freshPrivateName11193;`, and only lexically resolved references get renamed along with the declaration. This means that dynamic invocations can't access those declarations, and there cannot be a name conflict for any situation where two different classes/mixins/... reuse the same name for a `lexical` declaration, even in the same library.

**Comment by modulovalue** [comment]
> I tend to agree with lrhn that there is little justification for having a mechanism that prevents access to a declaration in the same library


I'm not trying to make an argument for a general mechanism for preventing access to a declaration in the same library, but to support not having to be forced to assign an identifier to the representation object when it isn't needed.

Anything that ends up being unused adds complexity and needs to be considered when one is trying to reason about the code at hand. I'm not proposing that we should be able to hide something, but that we should have the ability **not having to hide anything at all, because there's nothing to hide in the first place**.

---

I like the syntax that Lasse mentioned, that is:

```dart
// In addition to ...:
extension type Foo(int i) {
  // The representationDeclaration introduces an identifier 'i' into the scope of the body of Foo and `i` is bound to the representation object.
}

/// ... the following would also be supported:
extension type Foo(int) {
  // The representationDeclaration does not introduce any identifiers. The representation object can only be accessed via `this`.
}
```

The changes to the grammar are also straightforward:

```bnf
<representationDeclaration> ::=
  ('.' <identifierOrNew>)? '(' <metadata> <type> <identifier> ')'
```

```bnf
<representationDeclaration> ::=
  ('.' <identifierOrNew>)? '(' <metadata> <type> <identifier>? ')'
```

That is, the identifier of a representationDeclaration becomes optional.

---


I have provided at least three arguments for why I believe this to be a change that would make extension types more convenient to work with. 

To reiterate, forcing users to assign a name to the representation object would have the following downsides:

* a redundant name is introduced into the scope of Foo because the represention object can already be accessed via this.
* specifying an unreadable name for the represention object (e.g., `extension type Foo(int _$dontUseThisName) {}`) to discourage direct access of the representation object pollutes the scope of the extension type with an identifier that clearly is not meant to be there. And hacks like:

```dart
extension type DontTouchMyRepresentation(void _v)  {
  int get i => _v; // A value of type 'void' can't be returned from the function 'i' because it has a return type of 'int'.
}
```

would promote the use of [unsafe](https://github.com/dart-lang/language/issues/3365#issuecomment-1732284090) casts.

* a representionDeclaration with a non-optional identifier forces users to give the represention object a name, and it's unclear to me that, in general, clear and useful names do exist here (`value`, `self`, `_`?).

All three can be mitigated by supporting an optional <representationDeclaration> identifier.

@lrhn @eernstg in what ways do you disagree with these points?

**Comment by lrhn** [comment]
> @lrhn @eernstg in what ways do you disagree with these points?

> * a redundant name is introduced into the scope of Foo because the represention object can already be accessed via this.

The representation object can always be accessed via `this`, but that's typed at the extension type, not the representation type.

The only general way to go to the representation type, if you need that, is to cast using `as`, which, as you point out, is an unsafe operation. (I prefer to call it "error-prone", because "safety" is so overloaded.)
And if you don't need to, then it doesn't *matter* that the value can be accessed via `this`.

If the extension type is a subtype of the representation type, you can just up-cast. 
That's actually the most realistic situation where there is potentially no use for a representation object getter, unless you're shadowing some of the representation type members, and need to call them.

> * specifying an unreadable name for the represention object (e.g., `extension type Foo(int _$dontUseThisName) {}`) to discourage direct access of the representation object pollutes the scope of the extension type with an identifier that clearly is not meant to be there.

Yep. So don't do that.

The "unnecessary name in my namespace" argument is real. It used to be something I worried about too, but after actually trying to work with extension types, I learned to stopped worrying and love the representation object getter. 
It's there if I need it, and it's night invisible if I name it `_`. 
An identifier named `_` really doesn't pollute a namespace. It's there, but it's not getting in the way of anything actually useful, and it's not calling attention to itself. If anything, it's saying "don't use me" much clearer than "_$DontUseMe".

(I'd actually be likely to name it `_` even if I use it. It's a lovely short name.)

>  And hacks like:
> 
> ```dart
> extension type DontTouchMyRepresentation(void _v)  {
>   int get i => _v; // A value of type 'void' can't be returned from the function 'i' because it has a return type of 'int'.
> }
> ```
>
> would promote the use of [unsafe](https://github.com/dart-lang/language/issues/3365#issuecomment-1732284090) casts.

The `void` hack doesn't promote using casts. It's a hack, because it makes your representation type be `void`, which may not be what you want. I wouldn't use it.

But even if you do, it doesn't promote using casts, not any more than not having a representation object getter at all. 
If someone is willing to cast to get the representation object, they don't need no representation object getter, they can just cast the extension type'd value directly.
Having a representation object getter can only *discourage* casting, by providing the object at one more type without casting.

The getter does allow someone who *wouldn't* do a cast, to get access to the representation object at the representation type, something that would otherwise require casting, which they wouldn't do.

But this is code *inside your own library*, which would choose to use `foo._` to access the representation object.
So just don't write that code. It's your library.

The example makes no sense in this context, if you _want_ an `i` getter which accesses the representation object as an `int`, just write:
```dart
 extension type DontTouchMyRepresentation(int i)  {}
```
That's exactly what a public representation object getter is there for.


> * a representionDeclaration with a non-optional identifier forces users to give the represention object a name, and it's unclear to me that, in general, clear and useful names do exist here (`value`, `self`, `_`?).

If the getter is not to be used, which is what not having one would imply, the name `_` is fine. I'd use that.

I expect that it's going to be *incredibly rare* to have an extension type, which is not a subtype if its representation type, and which never needs to access the representation object at the representation type.
Possible, sure. But not common enough that it's going to matter.

> All three can be mitigated by supporting an optional <representationDeclaration> identifier.

Absolutely. 

As Erik said, we can "easily" define that `extension type Name(SomeType /*no identfier*/) ...` will not have a getter, or it will have a getter with a fresh private name, `_$87a934dc_6286_4733_9e84_b07bca21f27c`, that nobody is told about.

It does break our story, that an extension type is "like an inlined wrapper object" with the representation object getter as the only field. Or it doesn't, if we just say that it has a fresh name. Then everything is the same, there is a name polluting your namespace, it's just one you can't write by accident (or deliberately). But it's there.

It may break our story about the `Name(SomeType name)` being a "primary constructor", because primary constructor parameters like these introduce fields. But again, "freshly named" would at least make its behavior be explainable, even if other classes' primary constructors cannot omit names.

But it's possible.

It's not clear, to me, that even the marginal complexity of making the name optional, is worth the effect.

TL;DR:
* A representation object getter *cannot* promote casting. It can only provide a way to avoid a cast.
* Which means it does provide the representation object at a type that would otherwise require a cast.
* If you name the getter `_`, it really doesn't pollute the namespace. People are trained to ignore that name. 
* Don't consider any other hacks to discourage using the getter, they're not worth it, and they actually do add noise.
* Again, it's your own library. If you don't want to cast, don't. If you don't want to use the `_`-named getter, don't. If you do, do. Nobody else will ever need to know or care.
* Don't fret about the getter, or constructor, existing. Assume they exist whether you can name them or not.

**Comment by eernstg** [comment]
@modulovalue wrote:
> I have provided at least three arguments for why I believe this to be a change that would make extension types more convenient to work with.
>
> To reiterate, forcing users to assign a name to the representation object would have the following downsides:
>
> * a redundant name is introduced into the scope of Foo because the represention object can already be accessed via this.
> * specifying an unreadable name for the represention object (e.g., `extension type Foo(int _$dontUseThisName) {}`) to discourage direct access of the representation object pollutes the scope of the extension type with an identifier that clearly is not meant to be there. And hacks like:
>
> ```dart
> extension type DontTouchMyRepresentation(void _v)  {
>   int get i => _v; // A value of type 'void' can't be returned from the function 'i' because it has a return type of 'int'.
> }
> ```
>
> would promote the use of [unsafe](https://github.com/dart-lang/language/issues/3365#issuecomment-1732284090) casts.
>
> * a representionDeclaration with a non-optional identifier forces users to give the represention object a name, and it's unclear to me that, in general, clear and useful names do exist here (`value`, `self`, `_`?).

I didn't actually spot any downsides in that list.

I'm trying really hard to come up with a situation where we'd want an extension type `E` whose representation object is guaranteed to have a specific type `T`, and we never wish to use any members of `T`'s interface on that object (not in the body of `E`, not in code outside the declaration of `E`, not in other libraries). What are the members of `E` going to do? Return `this.toString()`? You can't do much else if you insist that you can't use the members of `T`.

OK, if you _do_ want access to the interface of `T` then you can use `(this as T)` (in practice, this could be `(this as RepresentationTypeWhichMightBeRatherVerbose<And, Have, TypeArguments>)`) in every location where you'd otherwise use the name of the representation object (for example: `it`). But why would you do that? It is error prone and verbose, and I just don't see how `(this as RepresentationTypeWhichMightBeRatherVerbose<And, Have, TypeArguments>).bar()` is better than `it.bar()`.

Hence, I don't think the name of the representation object is **redundant**: It does matter that the name of the representation object has the representation type, and `this` has the enclosing extension type.

About the **unreadable** name, and the pollution of the name space inside the extension type declaration: I'd expect `_` to work just fine, and I still don't see any use cases where that name is actually unused.

About using **the type `void`**: You wanted to be unable to access the representation object using the representation type, I just responded that having the type `void` is somewhat similar to being out of reach. If you want to do `_i as int` then _it is not true_ that you don't want to use the representation object with the representation type, and then you might as well give it a name.

About **choosing a clear and useful name** for the representation object: I think it would be perfectly fine if the community introduces some conventions about this choice. For example, they could choose to use `it` as the name of the representation object in every extension type. I think a certain amount of standardization would be good for the overall readability of extension types.

However, the fact that it's just a name may be helpful in the case where that very specific name which is chosen by convention must be the name of some other member of the extension type. In that case you just ignore the convention because you have to avoid the name clash, and use a different name for the representation object. No big deal, and you can still read the code as soon as you've noticed that it's slightly atypical.

**Comment by modulovalue** [comment]
Thank you @eernstg and @lrhn.

It's clear to me now that I had made a false assumption when it comes to the current design of extension types. I falsely assumed that `this` gives access to all members of the representation object + all members of the extension type, and that the extension type is not able to "statically override" any methods of the representation object. Both of these are incorrect.

As Lasse said, this makes it rather difficult to support a `Foo(int)` in addition to `Foo(int it)`. I agree with that.

And as Erik said, a new scoping mechanism could help with the issues that I've raised.

The premise here was that it would be rather simple to hide the representationDeclaration getter without sacrificing anything, but that is not the case. I'm going to close this issue as resolved.

**Comment by eernstg** [comment]
Aha, then I have a better understanding of your perspective!

The fact that `this` has the type of the enclosing extension type and the representation variable has the representation type allows us to keep those two interfaces apart (except for the 5 instance members of `Object` which are always "forwarded" to the representation object and hence works identically for both types).

The reason why you'd want to keep those two interfaces apart include the following:

- If you're using the extension type as a kind of `newtype` then the interface of the extension type should not include all the members of the interface of the representation type, you'd want to _replace_ the interface by some other interface. So you would use `extension type NotC(C it) { ... }` to specify a fresh set of members for `NotC`, and implement them in terms of invocations of members of `C` using expressions like `it.m()`. But you wouldn't want to provide `m` as a member of the interface of `NotC` because the whole point is that `NotC` allows us to use a `C` in a specific an restricted manner via that `NotC` interface.
- If you're using the extension type to change specific members then you'd redeclare those members (and possible allow direct forwarding of all the other members to the representation type).

For example, consider the case where we have a class whose interface isn't optimally typed (we've had that `update` method for many years, and we can't break the whole world by changing it to `void update(int j) => i = j;`):

```dart
// Legacy code.

class A {
  int i;
  A(this.i);
  void update(num n) => i = n as int;
  ... // Other members are OK.
}

// Modernized variant.
import 'legacy.dart' as legacy;

extension type A._(legacy.A it) implements legacy.A {
  A(int i) => A(legacy.A(i));
  void update(int j) => it.i = j;
}

// Usage 1.
import 'legacy.dart';

void main() {
  A a = A();
  a.update(1.5); // Throws at run time!
  a.otherMethod();
}

// Usage 2, a newer version of usage 1.
import 'modernized.dart';

void main() {
  A a = A();
  a.update(1.5); // Compile-time error!
  a.otherMethod(); // Still OK.
}
```

**Comment by modulovalue** [comment]
@eernstg Wouldn't a migration from such a "suboptimally typed" interface to an "optimally typed" interface via extension types only be "surface-level deep"? That is, if we call `update` via a level of indirection, the migrated extension type doesn't help us with preventing all update-related runtime exceptions.


```dart
void main() {
  () {
    // Usage 2, a newer version of usage 1.
    AModern a = AModern(0);
    a.update(1.5); // Compile-time error!
    indirectUpdate(a); // Throws at run time!
  }();
  () {
    // Usage 1.
    A a = A(0);
    a.update(1.5); // Throws at run time!
    indirectUpdate(a); // Throws at run time!
  }();
}

void indirectUpdate(A a) {
  a.update(1.5);
}

// Modernized variant.
extension type AModern._(A it) implements A {
  factory AModern(int i) => AModern._(A(i));
  void update(int j) => it.i = j;
}

class A {
  int i;
  A(this.i);
  void update(num n) => i = n as int;
  // ... Other members are OK.
}
```

Or is the goal here, in the example that you have mentioned, to add support for a gradual migration from "legacy A" to "modern A" without having to force all users of "legacy A" to migrate to "modern A" at the same time?

**Comment by eernstg** [comment]
The goal is simply to turn the run-time error into a compile-time error in as many situations as possible, using a device which isn't very disruptive. It's probably not going to be trivial at all, but this could be used as an intermediate phase where the end result is that there is only one class `A`, and it has the safe member signature.

During this process you could have a mixture of libraries using the safe API and libraries using the legacy API. The extension type `A` can then be used to access instances of `legacy.A` in one library _L1_ more safely than using `legacy.A` directly, and another library _L2_ might still use the type `legacy.A` everywhere. So we need to add some glue code to ensure that every `legacy.A` is given the type `A` when it is passed from _L2_ to _L1_, and the opposite glue would be used when objects are passed from _L1_ to _L2_.

In any case, it's likely to be very easy to do certain things (like using local variables of type `A` where we used to use the type `legacy.A`), and other things are less convenient. For example, it is usually harder to make changes to superinterfaces (`class B extends legacy.A ...` can't be changed to have the extension type as a superinterface).

So you're right, this is definitely an approach that starts out thin, and requires serious thought whenever it is being made thicker.
